clang-tools-extra/clang-tidy/bugprone/UnintendedCharOstreamOutputCheck.cpp

```diff
@@ -0,0 +1,69 @@
+//===--- UnintendedCharOstreamOutputCheck.cpp - clang-tidy ----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "UnintendedCharOstreamOutputCheck.h"
+#include "clang/AST/Type.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/ASTMatchers/ASTMatchers.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::bugprone {
+
+namespace {
+
+// check if the type is unsigned char or signed char
+AST_MATCHER(Type, isNumericChar) {
+  const auto *BT = dyn_cast<BuiltinType>(&Node);
+  if (BT == nullptr)
+    return false;
+  const BuiltinType::Kind K = BT->getKind();
+  return K == BuiltinType::UChar || K == BuiltinType::SChar;
+}
+
+// check if the type is char
+AST_MATCHER(Type, isChar) {
+  const auto *BT = dyn_cast<BuiltinType>(&Node);
+  if (BT == nullptr)
+    return false;
+  const BuiltinType::Kind K = BT->getKind();
+  return K == BuiltinType::Char_U || K == BuiltinType::Char_S;
+}
+
+} // namespace
+
+void UnintendedCharOstreamOutputCheck::registerMatchers(MatchFinder *Finder) {
+  auto BasicOstream =
+      cxxRecordDecl(hasName("::std::basic_ostream"),
+                    // only basic_ostream<char, Traits> has overload operator<<
+                    // with char / unsigned char / signed char
+                    classTemplateSpecializationDecl(
+                        hasTemplateArgument(0, refersToType(isChar()))));
+  Finder->addMatcher(
+      cxxOperatorCallExpr(
+          hasOverloadedOperatorName("<<"),
+          hasLHS(hasType(hasUnqualifiedDesugaredType(
+              recordType(hasDeclaration(cxxRecordDecl(
+                  anyOf(BasicOstream, isDerivedFrom(BasicOstream)))))))),
+          hasRHS(hasType(hasUnqualifiedDesugaredType(isNumericChar()))))
+          .bind("x"),
+      this);
+}
+
+void UnintendedCharOstreamOutputCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *Call = Result.Nodes.getNodeAs<CXXOperatorCallExpr>("x");
+  const Expr *Value = Call->getArg(1);
+  diag(Call->getOperatorLoc(),
+       "%0 passed to 'operator<<' outputs as character instead of integer. "
+       "cast to 'unsigned' to print numeric value or cast to 'char' to print "
+       "as character")
+      << Value->getType() << Value->getSourceRange();
```

Comment:
You could generate auto-fixes, easiest way is to insert "+" character before such, in that case it will be cast to int

Reply:
That's a fairly obscure solution, relying on implicit arithmetic promotions. It's banned in some coding guidelines (e.g. MISRA). Maybe just add an explicit `static_cast<int>`? Otherwise give the option to the user to choose what datatype to cast to, defaulting to `int`.

Reply:
The problem with `int` versus `unsigned int` is that in case you're on a platform where `sizeof(char) == sizeof(int)` then casting a value that happens to encode to a negative integer will result in the printing of a negative number. Not sure if this is something intentional.

(Nevertheless, I would prefer if the warning said `unsigned int` instead of just `unsigned`. It's non-trivial to know that `unsigned` alone is perfectly fine and actually means the same thing as `unsinged int`.)

Reply:
I think easiest is to just let the user decide what type they want to use, keeping some sane default (`unsigned int` is fine IMO). 

Some guidelines also ban the usage of built-in types like `int` and require you to use `std::int32_t` instead to be explicit. Having an option would allow the user to specify that.

Reply:
For me, it is more clearer to cast unsigned char to unsigned int and signed char to int. This can avoid that when converting unsigned char to int, people are not sure whether the result of 255 is -1 or 255.

Reply:
Given the range only suggesting `int` should be safe. Adding `unsigned` in case the underlying type would make it more clearer though.

As in other cases I think it should be coding guideline agnostic and suggest the least intrusive (e.g. `std::int32_t` requires the `<cstdint>` include). But let's not start that discussion again...

Reply:
> requires the <cstdint> include

Fair point. Perhaps just cast to `int` is the best compromise. It is also what unary plus does. If users need more flexibility we can add an option later.

Reply:
option is already added.